spatialos-sdk/src/worker/component.rs

```diff
@@ -1,35 +1,463 @@
-pub struct ComponentUpdate {
-    pub component_id: ComponentId,
-    pub schema_type: *mut Schema_ComponentUpdate,
+// A data structure which represents all known component types. Used to generate an array of vtables to pass
+// to the connection object.
+pub struct ComponentDatabase {
+    component_vtables: Vec<worker::Worker_ComponentVtable>,
+}
+
+impl ComponentDatabase {
+    pub fn new() -> Self {
+        ComponentDatabase {
+            component_vtables: Vec::new(),
+        }
+    }
+
+    pub fn add_component<C: Component>(mut self) -> Self {
+        self.component_vtables.push(create_component_vtable::<C>());
+        self
+    }
+
+    pub(crate) fn to_worker_sdk(&self) -> *const worker::Worker_ComponentVtable {
+        self.component_vtables.as_ptr()
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.component_vtables.len()
+    }
+}
+
+// An object which contains a reference counted instance to T.
+struct ClientHandle<T> {
+    data: Rc<T>,
+    // When Drop() is called, data should reduce its reference count.
 }
 
-impl From<&Worker_ComponentUpdate> for ComponentUpdate {
-    fn from(update: &Worker_ComponentUpdate) -> Self {
-        ComponentUpdate {
-            component_id: update.component_id,
-            schema_type: update.schema_type,
+impl<T> ClientHandle<T> {
+    fn new(data: T) -> ClientHandle<T> {
+        ClientHandle {
+            data: Rc::new(data),
+        }
+    }
+
+    fn copy(&self) -> ClientHandle<T> {
+        ClientHandle {
+            data: Rc::clone(&self.data),
         }
     }
+
+    // Vtable helper functions.
+    unsafe fn handle_allocate(data: T) -> *mut raw::c_void {
+        // Allocate client handle and initialize with data.
+        let new_client_handle: *mut ClientHandle<T> =
+            mem::transmute(alloc::alloc(Layout::new::<ClientHandle<T>>()));
+        ptr::write(new_client_handle, ClientHandle::<T>::new(data));
+
+        // Return new pointer.
+        mem::transmute(new_client_handle)
+    }
+
+    unsafe fn handle_free(handle: *mut raw::c_void) {
+        let client_handle_ptr: *mut ClientHandle<T> = mem::transmute(handle);
+
+        // Call drop() on pointer value.
+        client_handle_ptr.drop_in_place();
+
+        // Deallocate memory.
+        alloc::dealloc(
+            mem::transmute(client_handle_ptr),
+            Layout::new::<ClientHandle<T>>(),
+        );
+    }
+
+    unsafe fn handle_copy(handle: *mut raw::c_void) -> *mut raw::c_void {
+        let client_handle_ptr: *mut ClientHandle<T> = mem::transmute(handle);
+
+        // Allocate client handle and initialize with copy of existing client handle
+        let new_client_handle: *mut ClientHandle<T> =
+            mem::transmute(alloc::alloc(Layout::new::<ClientHandle<T>>()));
+        let client_handle: &ClientHandle<T> = &*client_handle_ptr;
+        ptr::write(new_client_handle, client_handle.copy());
+
+        // Return new pointer.
+        mem::transmute(new_client_handle)
```

Comment:
I think this function can be simplified down to:

```
let original = Arc::<T>::from_raw(handle);
let copy = original.clone();
mem::forget(original);
Arc::into_raw(copy) as *mut _
```

This will properly increment the counter for the `Arc` without

Reply:
Agree with this too.